features/cellular/framework/AT/AT_CellularDevice.cpp

```diff
@@ -318,6 +318,25 @@ void AT_CellularDevice::delete_context(CellularContext *context)
     release_at_handler(at);
 }
 
+nsapi_error_t AT_CellularDevice::restart()
+{
+    tr_debug("Restart cellular device and try to reconnect");
+
+    shutdown();
+    soft_power_off();
+    hard_power_off();
+
+    AT_CellularContext *curr = _context_list;
+    while (curr) {
+        nsapi_error_t error = curr->connect();
```

Comment:
Should check if a context has called connect and re-connection is really required.
In async mode should wait for `DISCONNECTED` events before calling `connect()`.

Reply:
@AriParkkila  How can we know if `connect()` was called before? Should I use `is_connected()` before shutting down the device?
Regarding the async mode I thought it will be handled through the `connect()` itself through the event queue. Can we change the mode temporary to synchronous mode here and then return it back after connecting? is there a better way for both points?